src/storage-types/src/read_holds.rs

```diff
@@ -0,0 +1,169 @@
+    pub fn new(
+        id: GlobalId,
+        since: Antichain<T>,
+        holds_tx: UnboundedSender<(GlobalId, ChangeBatch<T>)>,
+    ) -> Self {
+        Self {
+            id,
+            since,
+            holds_tx,
+        }
+    }
+
+    /// Returns the [GlobalId] of the collection that this [ReadHold] is for.
+    pub fn id(&self) -> &GlobalId {
+        &self.id
+    }
+
+    /// Returns the frontier at which this [ReadHold] is holding back the since
+    /// of the collection identified by `id`. This does not mean that the
+    /// overall since of the collection is what we report here. Only that it is
+    /// _at least_ held back to the reported frontier by this read hold.
+    pub fn since(&self) -> &Antichain<T> {
+        &self.since
+    }
+
+    /// Merges `other` into `self`, keeping the overall read hold.
+    pub fn merge_assign(&mut self, mut other: ReadHold<T>) {
+        let mut changes = ChangeBatch::new();
+
+        changes.extend(self.since.iter().map(|t| (t.clone(), -1)));
+
+        changes.extend(other.since.iter().map(|t| (t.clone(), -1)));
+        // It's very important that we clear the since of other. Otherwise, it's
+        // Drop impl would try and drop it again, by sending another ChangeBatch
+        // on drop.
+        let other_since = std::mem::take(&mut other.since);
+
+        self.since.extend(other_since.into_iter());
+
+        // Record the new requirements, which we're guaranteed to be possible
+        // because we're only retracing the two merged sinces together with this
+        // in one go.
+        changes.extend(self.since.iter().map(|t| (t.clone(), 1)));
+
+        match self.holds_tx.send((self.id.clone(), changes)) {
+            Ok(_) => (),
+            Err(e) => {
+                panic!("cannot merge ReadHold: {}", e);
+            }
+        }
+    }
+
+    /// Downgrades `self` to the given `frontier`. Returns `Err` when the new
+    /// frontier is `less_than` the frontier at which this [ReadHold] is
+    /// holding.
+    pub fn try_downgrade(&mut self, frontier: Antichain<T>) -> Result<(), anyhow::Error> {
+        if PartialOrder::less_than(&frontier, &self.since) {
+            return Err(anyhow::anyhow!(
+                "new frontier {:?} is beyond current since {:?}",
+                frontier,
+                self.since
+            ));
+        }
+
+        let mut changes = ChangeBatch::new();
+
+        changes.extend(self.since.iter().map(|t| (t.clone(), -1)));
+        changes.extend(frontier.iter().map(|t| (t.clone(), 1)));
+        self.since = frontier;
+
+        if !changes.is_empty() {
+            // If the other side already hung up, that's ok.
+            let _ = self.holds_tx.send((self.id.clone(), changes));
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: TimelyTimestamp> Clone for ReadHold<T> {
+    fn clone(&self) -> Self {
+        if self.id.is_user() {
+            tracing::debug!("cloning ReadHold on {}: {:?}", self.id, self.since);
```

Comment:
Do you want to keep this log in? Might want to consider using `trace!` instead, depending on how often this will fire in an idle environment that receives various metrics queries periodically.